### PR TITLE
reverseproxy: replace placeholders specified for sni while using http3

### LINKED ISSUE #7735: Dynamic http3 upstreams recieve untemplated sni
### Issue Details

Warning ahead, i have used AI to write the nix files to reproduce this issue. But confirmed that is not hallucination slop.

When i use a dynamic upstream using a wildcard host and http3 i recieve an untemplated sni on my host

example caddyfile
```
*.mydomain.com mydomain {
  reverse_proxy {
    dynamic a {
      name {host}
      port 443
      refresh 1m
      versions ipv6
    }
    transport http {
      tls
      tls_server_name {host}
      versions 3
    }
  }
}
```

this results in requests arriving at the upstream to get the SNI="{http.request.host}"

if i instead change the `tls_server_name {host}` to `tls_server_name git.mydomain.com` it works
if i instead remove `versions 3` it works

my upstream (also caddy) works on http3 as confirmed by `curl --http3-only git.mydomain.com`  on the same host that my proxy caddy lives on

this is also seen in the reproduction nix file output 

```
===== TSHARK (WORKING: H3 direct upstream (8081)) =====
        Interface name: any
        Interface name: any
                Extension: server_name (len=19) name=app.test.local
                    Type: server_name (0)
                    Server Name Indication extension
                        Server Name list length: 17
                        Server Name Type: host_name (0)
                        Server Name length: 14
                        Server Name: app.test.local
        Interface name: any
        Interface name: any
        Interface name: any
        Interface name: any
        Interface name: any
        Interface name: any
        Interface name: any
        Interface name: any
        Interface name: any
        Interface name: any
        Interface name: any
        Interface name: any
        Interface name: any
        Interface name: any
        Interface name: any
        Interface name: any
```

vs when calling via the wildcard dynamic http3 caddy reverse_proxy

```
===== TSHARK (BROKEN: H3 dynamic A upstream (8080)) =====
        Interface name: any
        Interface name: any
                Extension: server_name (len=24) name={http.request.host}
                    Type: server_name (0)
                    Server Name Indication extension
                        Server Name list length: 22
                        Server Name Type: host_name (0)
                        Server Name length: 19
                        Server Name: {http.request.host}
        Interface name: any
```

### Assistance Disclosure

AI used

### If AI was used, describe the extent to which it was used.

I wrote the nix file to reproduce this using ai

### PR DESCRIPTION
**Pull request by WeidiDeng**
Fix #7735 

## Assistance Disclosure

_No AI is used._

### CODE REVIEW

**Comment by steadytao** [comment]
Thanks Weidi. Confirmed working by reporter. The implementation looks narrowly scoped and I do not see an obvious unintended side effect for non-H3 or static-SNI upstreams. The main thing I would like before merge is a regression test for the actual bug, HTTP/3 upstream with `tls_server_name {host}` should expand the SNI per request instead of sending the literal placeholder. This path now uses a *technically* custom QUIC dial function so having coverage would make this much safer to carry.

**Comment by mholt** [comment]
Thank you Weidi! Yeah, I agree a test could be uesful here.

**Comment by WeidiDeng** [comment]
Actually, to make tests work, there is more to be done here.

http3 transport doesn't use the result from configured dns resolver and will always use system dns. To make that work, a custom dial function is a must. And to test the placeholder works, we need to create some kinds of config that will serve custom domains and instruct caddy to use the custom dial address.

**Comment by steadytao** [comment]
That makes sense. The test needs to decouple the logical upstream name/SNI from the actual local QUIC dial address, otherwise `http3.Transport` falls back to system DNS and cannot reliably route a fake test hostname to the local test server.

The shape would likely have to be something similar to:
- client request Host: app.test.local
- Caddy reverse_proxy uses HTTP/3 upstream transport
- transport has tls_server_name {host}
- custom H3 dial sends packets to local test QUIC server
- test server records ClientHello SNI
- assert SNI == app.test.local
- assert SNI != {http.request.host}

**Comment by WeidiDeng** [comment]
Perhaps you misunderstood. What I meant is that to test the placeholder issue is fixed, we need also fix the issue that custom DNS resolution is not used regardless if placeholder is used. 

Otherwise, to only use custom DNS resolution only if placeholder for sni is used is another bug that needs to be solved.

**Comment by steadytao** [comment]
I think we can and likely should keep the two issues decoupled.

The broader issue that HTTP/3 upstream dialing does not honour the configured resolver is real, but this PR does not need to solve it just to test placeholder SNI expansion. For this regression, the test can avoid DNS entirely by using a literal local upstream dial address and a separate logical request host, e.g. the test shape I mentioned earlier.

I would prefer that even if resolver support were already fixed, because the bug under test is placeholder SNI expansion for HTTP/3, not name resolution; keeping DNS out of the test means a failure points directly at the SNI placeholder path rather than the surrounding plumbing.

A separate PR can then make HTTP/3 upstream dialing honour the configured resolver consistently for both static and placeholder SNI if we deem necessary.

**Comment by WeidiDeng** [comment]
>I would prefer that even if resolver support were already fixed, because the bug under test is placeholder SNI expansion for HTTP/3, not name resolution; keeping DNS out of the test means a failure points directly at the SNI placeholder path rather than the surrounding plumbing.

I never said anything about not adding tests for placeholders for server name. Without custom address resolution, very few hosts are candidates for testing with hosts hack, like `localhost` and `127.0.0.1`.

>A separate PR can then make HTTP/3 upstream dialing honour the configured resolver consistently for both static and placeholder SNI if we deem necessary.

h3 transport being experimental doesn't mean bugs shouldn't be addressed. It's necessary, just not enough bandwidth can be devoted.

**Comment by WeidiDeng** [comment]
>I think we can and likely should keep the two issues decoupled.

Should be done doesn't mean it can be done. Right now there is no way to specify the static dial address for http3 as it will ignored as system dns is used. Fixing that more or less fixes custom dns resolution anyway.

**Comment by steadytao** [comment]
I think we may still be talking past each other. I agree that HTTP/3 resolver support is a real bug and I am not saying HTTP/3 bugs should be ignored. My point is **only** about the scope of this PRs regression test. I also agree that if we want `transport http { resolver ... }` to work for HTTP/3 upstreams, that likely needs changes to the HTTP/3 dial path. I am not claiming that is trivial or already possible in our current implementation.

The narrower point is that this PRs test does not need to prove custom resolver behaviour. It only needs to prove that placeholder SNI is expanded before the HTTP/3 upstream handshake. That can be tested with a literal loopback upstream address and a separate request Host.

That avoids DNS and keeps this regression focused on the behaviour changed by this PR. I am trying to avoid making one regression test cover two independent bugs, not arguing that the resolver bug is unimportant.

**Comment by steadytao** [comment]
The added test is not quite the regression I was asking for. It uses `tls_server_name {header.X-SNI}` and proxies back into Caddy’s own HTTPS listener so it covers a broader and more fragile path than necessary.

The bug to lock down is HTTP/3 upstream SNI expansion for `tls_server_name {host}`. I would prefer a test with a literal local upstream dial address, a separate request `Host` and a local H3 server that records ClientHello SNI then asserts SNI is the request host and not the literal placeholder.

**Comment by WeidiDeng** [comment]
This patch fixes the placeholder, it doesn't need to be `tls_server_name {host}`. Setting up another h3 server is too much work with associated tls certificates, and a lot other related certificate trust pool for the reverse proxy transport, for too little benefit. And that's easy to say to test the server receives the intended clienthello when clearly the correct response is enough.

> a literal local upstream dial address

As I said, using the custom dial address requires another fix which you said you do not want. If you think it's possible, you can add it in the patch or create a new patch to address it. I won't spend any more time on this.

**Comment by WeidiDeng** [comment]
This is where the custom dial address and dns resolution takes place for h1/2,

https://github.com/caddyserver/caddy/blob/44b667a79f48e6163570cd6b32fa806e12625516/modules/caddyhttp/reverseproxy/httptransport.go#L283-L290

Feel free to implement your own suggested change yourself without similar logic.

**Comment by WeidiDeng** [comment]
> The bug to lock down is HTTP/3 upstream SNI expansion for tls_server_name {host}

The bug is the placeholder not properly replaced, not the placeholder used for `tls_server_name`.

>a separate request Host

`Host` has nothing to do with `tls_server_name`. It is good to make sure `tls_server_name` match with `host` to prevent request smuggling, and caddy by default doesn't [requires this](https://caddyserver.com/docs/caddyfile/options#strict-sni-host) unless client authentication is used.

> Caddy’s own HTTPS listener so it covers a broader and more fragile path than necessary

So caddy's handling of http3 in this case is fragile and flawed? And the placeholder used to get some tls related information can't be trusted, and it's better to run an http3 server to make sure?

**Comment by WeidiDeng** [comment]
The point of tests is to make something behaves correctly in certain circumstances, not blindly replicating behaviors of users. Even for edge cases a test needn't to be created to fully match the decisions made by users, test needs to capture the core of the problem.

**Comment by steadytao** [comment]
I pushed a test update to make the requested regression coverage explicit. Weidi, to be clear, I agree with your point that tests should capture the core problem rather than blindly reproduce a users full setup. That is precisely why I wanted this shape.

The core bug in #7735 is not the dynamic A upstream, wildcard site block, Nix repro or DNS behaviour. It is this data flow:
- `request Host` -> `{host}` -> `tls_server_name` -> outbound HTTP/3 ClientHello SNI

The updated test isolates that path. It uses a literal loopback HTTP/3 upstream address, sends the client request with `Host: app.test.local`, configures `tls_server_name {host}` and has a local H3 server record `tls.ClientHelloInfo.ServerName`.

That is intentional. The bug is about the SNI value sent by Caddy during the outbound HTTP/3 handshake so the strongest assertion point is the receiving H3 server’s ClientHello view. This avoids routing the check through Caddy’s own inbound TLS/request placeholder path and it avoids DNS so this does not depend on or attempt to solve the separate HTTP/3 resolver issue.

Please review when you have time.

**Comment by mholt** [comment]
Ok, finally read through this and I have some idea of what is going on... sorry for my slowness. Holiday weekend :)

Basically, the main contention is over the test approach. Weidi's test is simple, and uses a a placeholder of an HTTP header, and asserts proper expansion/replacement by responding with the ServerName presented in the handshake to the backend. Steady's test is more complex but directly asserts the ServerName by inspecting it and asserting against it during the TLS handshake on the backend.

Weidi's test relies on other parts of Caddy working properly (i.e. parts of the Caddyfile implement logic used by test assertions), and Steady's test is a more direct inspection that does not rely on Caddy conveying a proper HTTP response from the backend to the proxy.

I see value in both approaches. I like the simplicity of Weidi's approach and most of those other parts of Caddy are under test so I view it as reliable. I also like Steady's approach for it being such a direct inspection -- more "pure" perhaps -- but I don't know if the extra code is worth it. (It's a lot of code for simply inspecting a string value.) I think I'm leaning toward Weidi's test, honestly? Thoughts?

As an aside, the patch itself makes sense to me now too, but I do think it's unfortunate that we need to copy so much of quic-go's internal guts and store a separate quicTransport just to expand placeholders in the dial. Would there be an exported API we could propose to @marten-seemann to make this easier and more maintainable?

Thank you both @WeidiDeng and @steadytao ... really appreciate your patience on this.

**Comment by steadytao** [comment]
That matches my concern. I preferred the direct H3 server test because the bug is specifically the outbound ClientHello SNI value and this code path is already a bit unusual since we are replacing part of quic-go’s dial path to mutate SNI before the H3 handshake.

I agree it is more code for one assertion but most of that is isolated H3/cert setup and the actual test body is small. That said, I am fine keeping the simpler test if you think it is enough. I would just prefer it use `tls_server_name {host}` if practical since that is the path from #7735.

Agreed on quic-go too. The copied dial/resolver code is my largest long-term concern.

File: caddytest/integration/reverseproxy_test.go [hunk 1]
```diff
@@ -793,3 +802,140 @@ func TestReverseProxyRetryMatchIsTransportError(t *testing.T) {
...
+		case err := <-errs:
+			t.Errorf("HTTP/3 upstream server failed: %v", err)
+		default:
+		}
+	})
+
+	return udpConn.LocalAddr().String(), gotSNI
+}
+
+func http3SNITestTLSConfig(t *testing.T, recordSNI func(string)) *tls.Config {
```
Comment (WeidiDeng) [inline caddytest/integration/reverseproxy_test.go]: If tests are to be added this way, this can also be used for the (yet) non-existent tests for h1/h2 transports for the same `tls_server_name` placeholder with minimal changes.

File: caddytest/integration/reverseproxy_test.go [hunk 2]
```diff
@@ -793,3 +802,140 @@ func TestReverseProxyRetryMatchIsTransportError(t *testing.T) {
...
+	tester.AssertResponse(req, 200, "ok")
+
+	select {
+	case sni := <-gotSNI:
+		if sni != expectedSNI {
+			t.Fatalf("HTTP/3 upstream SNI = %q, want %q", sni, expectedSNI)
+		}
+		if sni == "{http.request.host}" {
+			t.Fatal("HTTP/3 upstream SNI was not expanded from the adapted placeholder")
+		}
```
Comment (WeidiDeng) [inline caddytest/integration/reverseproxy_test.go]: Both checks exclude each other. Only one should be kept.

**Comment by mholt** [comment]
Personally I prefer the simpler test (Weidi's test) -- unless this more complex test harness can be updated to cover more behavior.

Would either of you have an API to propose for Marten that could make it so we don't have to copy quic-go's internals?

**Comment by steadytao** [comment]
If the heavier harness stays, I agree it should probably cover H1/H2/H3 placeholder SNI to justify the extra setup.

For quic-go, I would ask about a pre-dial TLS config hook on `http3.Transport`, called after cloning `TLSClientConfig` but before default dial/resolution. Then Caddy could mutate `cfg.ServerName` without replacing `Dial` and copying quic-go internals.

**Comment by mholt** [comment]
(Didn't Weidi's tests also cover the 1.1 and 2 placeholders?)

**Comment by steadytao** [comment]
Yes, you’re right. I’m fine keeping Weidi’s simpler test if that is the preferred tradeoff.

**Comment by mholt** [comment]
Thank you for the iterative discussion and patch!
